pkg/relayer/session/session.go

```diff
@@ -0,0 +1,123 @@
+package session
+
+import (
+	"context"
+	blockclient "pocket/pkg/client"
+	"pocket/pkg/observable"
+	"pocket/pkg/observable/channel"
+	sessiontypes "pocket/x/session/types"
+	"sync"
+)
+
+var _ RelayerSessions = (*relayerSessions)(nil)
+
+type (
+	sessionId        = string
+	blockHeight      = int64
+	sessionsTreesMap = map[blockHeight]map[sessionId]SessionTree
+)
+
+// relayerSessions is an implementation of the RelayerSessions interface.
+type relayerSessions struct {
+	// closingSessions notifies about sessions that are ready to be claimed.
+	closingSessions observable.Observable[SessionTree]
+
+	// closingSessionsPublisher is the channel used to publish closing sessions.
+	closingSessionsPublisher chan<- SessionTree
+
+	// sessionTrees is a map of block heights containing an other map of SessionTrees indexed by their sessionId.
+	sessionsTrees   sessionsTreesMap
+	sessionsTreesMu *sync.Mutex
+
+	// blockClient is the block client used to get the committed blocks notifications.
+	blockClient blockclient.BlockClient
+
+	// storesDirectory is the directory where the KVStore data files are created.
+	storesDirectory string
+}
+
+// NewRelayerSessions creates a new relayerSessions.
+func NewRelayerSessions(
+	ctx context.Context,
+	storesDirectory string,
+	blockClient blockclient.BlockClient,
+) RelayerSessions {
+	rs := &relayerSessions{
+		sessionsTrees:   make(sessionsTreesMap),
+		storesDirectory: storesDirectory,
+		blockClient:     blockClient,
+	}
+	rs.closingSessions, rs.closingSessionsPublisher = channel.NewObservable[SessionTree]()
+
+	go rs.goListenToCommittedBlocks(ctx)
+
+	return rs
+}
+
+// ClosingSessions returns an observable that notifies when sessions are ready to be claimed.
+func (rs *relayerSessions) ClosingSessions() observable.Observable[SessionTree] {
+	return rs.closingSessions
+}
+
+// EnsureSessionTree returns the SessionTree for a given session.
+// If the session is encountered for the first time, a SessionTree is created for it before returning.
+func (rs *relayerSessions) EnsureSessionTree(session *sessiontypes.Session) (SessionTree, error) {
+	rs.sessionsTreesMu.Lock()
+	defer rs.sessionsTreesMu.Unlock()
+
+	// Calculate the session end height based on the session start block height
+	// and the number of blocks per session.
+	sessionEndHeight := session.Header.SessionStartBlockHeight + session.NumBlocksPerSession
```

Comment:
Since we are looking for sessions that end at the current block height, it seemed to be easier to calculate the sessionEndHeight and compare it to the current block height.

If we use the session number, we would have to convert the current block number to a sessionNumber each time there is a new block, then check for sessions that live in the `sessionNumber - 1` map key which feels a bit backward.

WDYT?

Reply:
I think it's reasonable. Please add a comment next to `sessionTrees` to explain this in more detail though.